File: supabase/functions/analyze-food/scoring.ts
import { NutritionInfo } from "./types.ts";
import { ensureModelTrained } from "./model.ts";

export function toNumber(value: unknown, fallback: number): number {
  const n = typeof value === 'number' ? value : typeof value === 'string' ? parseFloat(value) : NaN;
  return Number.isFinite(n) ? n : fallback;
}

// Estimate processing level from product data
export function getProcessingLevel(product: any): number {
  const ingredients = product.ingredients_text?.toLowerCase() || '';
  const categories = product.categories?.toLowerCase() || '';
  
  let score = 0.5;
  
  if (ingredients.includes('preservative') || ingredients.includes('artificial')) score += 0.2;
  if (ingredients.includes('color') || ingredients.includes('flavoring')) score += 0.15;
  if (categories.includes('fresh') || categories.includes('raw')) score -= 0.3;
  if (categories.includes('organic')) score -= 0.2;
  if (categories.includes('processed') || categories.includes('ultra-processed')) score += 0.3;
  
  return Math.max(0.1, Math.min(1.0, score));
}

/**
 * Goal-specific feature weight multipliers based on user targetBodyType.
 * Adjusts the linear regression weights and feature contributions to align with individual fitness goals:
 * 
 * Multiplier Matrix:
 * Feature index order: [Calories, Protein, Carbs, Fat, SatFat, Sugar, Fiber, Sodium, Processing]
 * - athletic:     [0.9,      1.4,     1.0,   1.0, 1.1,    1.1,   1.2,   1.0,    1.0] -> Upweights protein (1.4x) & fiber (1.2x); allows moderate calories for muscle growth.
 * - muscle_gain:  [0.9,      1.4,     1.0,   1.0, 1.1,    1.1,   1.2,   1.0,    1.0] -> Synonym for athletic.
 * - thin:         [1.3,      1.1,     1.1,   1.1, 1.2,    1.3,   1.3,   1.1,    1.0] -> Emphasizes calorie/sugar restriction and fiber for lean definition.
 * - weight_loss:  [1.3,      1.1,     1.1,   1.1, 1.2,    1.3,   1.3,   1.1,    1.0] -> Synonym for thin/fat-loss.
 * - overweight:   [1.4,      1.1,     1.1,   1.2, 1.3,    1.4,   1.3,   1.2,    1.0] -> Stronger calorie/sugar/fat penalty to support metabolic health.
 * - obese:        [1.4,      1.1,     1.1,   1.2, 1.3,    1.4,   1.3,   1.2,    1.0] -> Strongest calorie/sugar/fat penalty.
 * - average:      [1.0,      1.0,     1.0,   1.0, 1.0,    1.0,   1.0,   1.0,    1.0] -> Base model weights unchanged (maintenance).
 * - maintenance:  [1.0,      1.0,     1.0,   1.0, 1.0,    1.0,   1.0,   1.0,    1.0] -> Synonym for average.
 */
export const GOAL_FEATURE_MULTIPLIERS: Record<string, number[]> = {
  athletic:    [0.9, 1.4, 1.0, 1.0, 1.1, 1.1, 1.2, 1.0, 1.0],
  muscle_gain: [0.9, 1.4, 1.0, 1.0, 1.1, 1.1, 1.2, 1.0, 1.0],
  thin:        [1.3, 1.1, 1.1, 1.1, 1.2, 1.3, 1.3, 1.1, 1.0],
  weight_loss: [1.3, 1.1, 1.1, 1.1, 1.2, 1.3, 1.3, 1.1, 1.0],
  overweight:  [1.4, 1.1, 1.1, 1.2, 1.3, 1.4, 1.3, 1.2, 1.0],
  obese:       [1.4, 1.1, 1.1, 1.2, 1.3, 1.4, 1.3, 1.2, 1.0],
  average:     [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
  maintenance: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
};

export function getGoalMultipliers(targetBodyType?: string): number[] {
  const key = (targetBodyType || 'athletic').toLowerCase();
  return GOAL_FEATURE_MULTIPLIERS[key] || GOAL_FEATURE_MULTIPLIERS.athletic;
}

// Calculate basic health score for training labels & rule-based fallback
export function calculateBasicHealthScore(nutriments: any, targetBodyType: string = 'athletic'): number {
  const calories = toNumber(nutriments['energy-kcal_100g'] ?? (nutriments.energy_100g ? nutriments.energy_100g / 4.184 : 0) ?? nutriments.calories, 0);
  const protein = toNumber(nutriments.proteins_100g ?? nutriments.protein, 0);
  const fiber = toNumber(nutriments.fiber_100g ?? nutriments.fiber, 0);
  const fat = toNumber(nutriments.fat_100g ?? nutriments.fat, 0);
  const saturatedFat = toNumber(nutriments['saturated-fat_100g'] ?? nutriments.saturatedFat, 0);
  const sugar = toNumber(nutriments.sugars_100g ?? nutriments.sugar, 0);
  
  // Note: Open Food Facts sodium_100g is in grams; if < 20 and > 0, convert to mg. Otherwise assume mg.
  const rawSodium = toNumber(nutriments.sodium_100g ?? nutriments.sodium, 0);
  const sodium = rawSodium < 20 && rawSodium > 0 ? rawSodium * 1000 : rawSodium;
  
  const mult = getGoalMultipliers(targetBodyType);
  let score = 50;
  
  // Positive factors (adjusted by goal multipliers)
  if (protein > 10) score += 15 * mult[1];
  if (fiber > 5) score += 15 * mult[6];
  
  // Negative factors (adjusted by goal multipliers)
  if (calories > 400) score -= 15 * mult[0];
  if (fat > 20) score -= 10 * mult[3];
  if (saturatedFat > 10) score -= 10 * mult[4];
  if (sugar > 20) score -= 15 * mult[5];
  if (sodium > 500) score -= 10 * mult[7];
  
  return Math.max(0, Math.min(100, score));
}

// Calculate health score using trained ML model or fallback with goal-specific weighting
export async function calculateHealthScore(nutrition: NutritionInfo, targetBodyType: string = 'athletic'): Promise<number> {
  const model = await ensureModelTrained();
  const multipliers = getGoalMultipliers(targetBodyType);
  
  if (model && model.weights && model.weights.length >= 9) {
    // Features normalized for scoring
    const features = [
      (nutrition.calories || 0) / 600,
      (nutrition.protein || 0) / 30,
      (nutrition.carbs || 0) / 100,
      (nutrition.fat || 0) / 40,
      (nutrition.saturatedFat || 0) / 20,
      (nutrition.sugar || 0) / 50,
      (nutrition.fiber || 0) / 15,
      (nutrition.sodium || 0) / 2000,
      nutrition.processingLevel ?? 0.5
    ];
    
    // Apply goal multipliers to weights for target-specific scoring
    let result = model.bias;
    for (let i = 0; i < features.length; i++) {
      const effectiveWeight = model.weights[i] * (multipliers[i] ?? 1.0);
      result += effectiveWeight * features[i];
    }
    const score = (1 / (1 + Math.exp(-result))) * 100;
    return Math.max(0, Math.min(100, score));
  }
  
  // Fallback to rule-based scoring with goal adjustments if model not available
  return calculateBasicHealthScore({
    'energy-kcal_100g': nutrition.calories,
    'proteins_100g': nutrition.protein,
    'fat_100g': nutrition.fat,
    'saturated-fat_100g': nutrition.saturatedFat,
    'sugars_100g': nutrition.sugar,
    'fiber_100g': nutrition.fiber,
    'sodium': nutrition.sodium
  }, targetBodyType);
}

File: python_app/scoring.py
from typing import Dict, Any, List, Optional
from .models import NutritionInfo

def to_number(value: Any, fallback: float) -> float:
    if value is None:
        return fallback
    if isinstance(value, (int, float)):
        return float(value)
    try:
        return float(str(value).strip())
    except (ValueError, TypeError):
        return fallback

def get_processing_level(product: Dict[str, Any]) -> float:
    ingredients = str(product.get("ingredients_text", "")).lower()
    categories = str(product.get("categories", "")).lower()
    
    score = 0.5
    if "preservative" in ingredients or "artificial" in ingredients:
        score += 0.2
    if "color" in ingredients or "flavoring" in ingredients:
        score += 0.15
    if "fresh" in categories or "raw" in categories:
        score -= 0.3
    if "organic" in categories:
        score -= 0.2
    if "processed" in categories or "ultra-processed" in categories:
        score += 0.3
    
    return max(0.1, min(1.0, score))

GOAL_FEATURE_MULTIPLIERS: Dict[str, List[float]] = {
    "athletic":    [0.9, 1.4, 1.0, 1.0, 1.1, 1.1, 1.2, 1.0, 1.0],
    "muscle_gain": [0.9, 1.4, 1.0, 1.0, 1.1, 1.1, 1.2, 1.0, 1.0],
    "thin":        [1.3, 1.1, 1.1, 1.1, 1.2, 1.3, 1.3, 1.1, 1.0],
    "weight_loss": [1.3, 1.1, 1.1, 1.1, 1.2, 1.3, 1.3, 1.1, 1.0],
    "overweight":  [1.4, 1.1, 1.1, 1.2, 1.3, 1.4, 1.3, 1.2, 1.0],
    "obese":       [1.4, 1.1, 1.1, 1.2, 1.3, 1.4, 1.3, 1.2, 1.0],
    "average":     [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
    "maintenance": [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
}

def get_goal_multipliers(target_body_type: Optional[str] = "athletic") -> List[float]:
    key = (target_body_type or "athletic").lower()
    return GOAL_FEATURE_MULTIPLIERS.get(key, GOAL_FEATURE_MULTIPLIERS["athletic"])

def calculate_basic_health_score(nutriments: Dict[str, Any], target_body_type: str = "athletic") -> float:
    calories = to_number(nutriments.get("energy-kcal_100g") or (float(nutriments.get("energy_100g", 0)) / 4.184 if nutriments.get("energy_100g") else None) or nutriments.get("calories"), 0.0)
    protein = to_number(nutriments.get("proteins_100g") or nutriments.get("protein"), 0.0)
    fiber = to_number(nutriments.get("fiber_100g") or nutriments.get("fiber"), 0.0)
    fat = to_number(nutriments.get("fat_100g") or nutriments.get("fat"), 0.0)
    saturated_fat = to_number(nutriments.get("saturated-fat_100g") or nutriments.get("saturatedFat"), 0.0)
    sugar = to_number(nutriments.get("sugars_100g") or nutriments.get("sugar"), 0.0)
    processing_level = to_number(nutriments.get("processingLevel"), 0.3)
    
    raw_sodium = to_number(nutriments.get("sodium_100g") or nutriments.get("sodium"), 0.0)
    sodium = raw_sodium * 1000 if 0 < raw_sodium < 20 else raw_sodium
    
    mult = get_goal_multipliers(target_body_type)
    score = 55.0
    
    # Positive factors
    score += min(25.0, protein * 1.5) * mult[1]
    score += min(20.0, fiber * 2.5) * mult[6]
    
    if processing_level <= 0.3:
        score += 15.0 * mult[8]
    elif processing_level > 0.6:
        score -= 15.0 * mult[8]
        
    if calories < 120 and sugar < 12 and saturated_fat < 1:
        score += 12.0
        
    # Negative factors
    if calories > 350:
        score -= min(25.0, (calories - 350) / 15.0) * mult[0]
    if fat > 15:
        score -= min(20.0, (fat - 15) * 1.0) * mult[3]
    if saturated_fat > 3:
        score -= min(20.0, (saturated_fat - 3) * 2.0) * mult[4]
    if sugar > 12:
        score -= min(25.0, (sugar - 12) * 1.0) * mult[5]
    if sodium > 300:
        score -= min(15.0, (sodium - 300) / 50.0) * mult[7]
        
    return max(5.0, min(98.0, round(score, 1)))

async def calculate_health_score(nutrition: NutritionInfo, target_body_type: str = "athletic") -> float:
    from .model import ensure_model_trained
    model = await ensure_model_trained()
    multipliers = get_goal_multipliers(target_body_type)
    
    if model and len(model.weights) >= 9:
        features = [
            (nutrition.calories or 0.0) / 600.0,
            (nutrition.protein or 0.0) / 30.0,
            (nutrition.carbs or 0.0) / 100.0,
            (nutrition.fat or 0.0) / 40.0,
            (nutrition.saturatedFat or 0.0) / 20.0,
            (nutrition.sugar or 0.0) / 50.0,
            (nutrition.fiber or 0.0) / 15.0,
            (nutrition.sodium or 0.0) / 2000.0,
            nutrition.processingLevel if nutrition.processingLevel is not None else 0.5
        ]
        
        score = model.bias
        for i in range(len(features)):
            effective_weight = model.weights[i] * multipliers[i]
            score += effective_weight * features[i]
            
        return max(5.0, min(98.0, round(score, 1)))
        
    return calculate_basic_health_score({
        "calories": nutrition.calories,
        "protein": nutrition.protein,
        "fat": nutrition.fat,
        "saturatedFat": nutrition.saturatedFat,
        "sugar": nutrition.sugar,
        "fiber": nutrition.fiber,
        "sodium": nutrition.sodium,
        "processingLevel": nutrition.processingLevel
    }, target_body_type)
